### PR TITLE
[Ready] Communal Resource Storage II

### LINKED ISSUE #39232: Medical TechFab will not accept plastic, so no XL beakers
As simple as the title says, the Medical TechFab wont accept any plastic

Current Test Merges
#39118: '[Ready] Communal Resource Storage II' by SpaceManiac at commit 6dc4365c52
#39136: 'Removes can_vv_get from some things' by Cyberboss at commit 75ea5e3c1b
#39220: 'Makes having high body temperature more deadly in general' by vuonojenmustaturska at commit 98cfe0e111

Personally I believe it has something to do with #39118 because the ORM being magically linked to TechFab, and ORM has no spot for plastic.

Wanted to let you know because I cant make my juicy XL beakers

### PR DESCRIPTION
**Pull request by SpaceManiac**
[Sponsored](https://tgstation13.org/phpBB/viewtopic.php?f=5&t=17661) by @optimumtact. Based in part on #37901.

:cl:
add: The vault now contains an ore silo where the station's minerals are stored.
add: The station's ORM, recycling, and the labor camp send materials to the silo via bluespace.
add: Protolathes, techfabs, and circuit imprinters all pull materials from the silo via bluespace.
add: Those with vault access can view mineral logs and pause or remove any machine's access, or add machines with a multitool.
tweak: The ORM's alloy recipes are now available in engineering and science protolathes.
/:cl:

* [x] Fix #39232
  * [x] by allowing the silo to store plastic?
  * ~~biomass?~~
* [x] Componentize the remote material storage (current structure has too much duplication)
* ~~Consider adding exosuit fabricators to the linkage~~
  * Not in this PR, but if someone else wants to add this later, I can advise
* ~~Consider hiding exactly what is being printed (is this an issue in practice?)~~
* ~~Consider reducing ability to eject all sheets at a protolathe~~
  * ~~(?) Possibly return local storage and have the silo apportion materials~~
* [x] Handle the general case of off-station lathes rather than special-casing the golem ORM
  * [x] Give non-siloed protolathes limited storage again
* [x] Paginate ore silo logs
* [x] Make the ore silo constructible

----

* [x] Add ore silo machine
  * [x] Write a description
  * ~~Finalize sprite and effects~~
    * MrDoomBringer will be adding a sprite in a future PR
* [x] Mapping changes
  * [x] Place ore silo in vault
  * ~~Combine vault into Cargo warehouse~~
    * Deferred to avoid interfering with #39177 or any other remap which would leave something useful where the vault used to be
* [x] Retrieve materials from silo for R&D fabs
* [x] Remove material bins from R&D fab component parts
* [x] Send smelted materials from ORM to silo
  * [x] Refuse to smelt if there is no silo
  * ~~Remove ability to retrieve sheets from ORM~~
* [x] Send reclaimed materials from DA to silo
* [x] Send inserted materials from R&D fabs to silo
* [x] Send materials from stacking machines to silo
  * [x] Labor shuttle
  * [x] (?) Recycling
* [x] Automatically link mapload R&D fabs to silo
* [x] Allow linking machines to silo with multitool
* [x] Allow unlinking machines from silo UI
* [x] Allow putting access on hold without unlinking fabs
* [x] Log silo deposits and withdrawals, show in UI
  * [x] View log from QM PDA
* [x] Update R&D fab logic to handle disconnected silo
* [x] Update R&D fab UI to either
  * ~~disallow ejecting materials, or~~
  * [x] eject materials locally rather than at the silo
* [x] Allow ejecting materials from the silo UI
* [x] Add smelting designs from ORM to protolathes
  * [x] Finalize which departments get these recipes
* [x] Free golem ORM should either
  * ~~have its own silo, or~~
  * [x] not use silos, retaining old behavior

image

image

### CODE REVIEW

**Comment by Iamgoofball** [comment]
I'd advise just disallow ejecting and just let people "print" sheets.

**Comment by Wilchenx** [comment]
image
This is bad if silo gets bombed/broken/stolen/depowered(depowering does turn off ability to siphon minerals, right?). ORM should retain ability to smelt ores as a backup if it cannot see a silo.

**Comment by Iamgoofball** [comment]
@Wilchenx aren't you banned still

**Comment by Wilchenx** [comment]
@Iamgoofball off-topic

**Comment by davethwave** [comment]
I would say that the golem ORM should not have a silo since there's no reason for it.

**Comment by kevinz000** [comment]
IMO ORM should retain the ability to be the default ORM if the person smelting specifically hits the button for it, but default to sending to silo.

**Comment by Grazyn** [comment]
Is the golem ORM different from the other omnilathes present in the game? It would be nice to be able to link the omnilathe to the silo if you manage to find the boards and bring them to the station

**Comment by Grazyn** [comment]
Wait can't golem just make a department protolathe and use it to siphon minerals? Maybe it's better to make the silo not work across z-levels

**Comment by Grazyn** [comment]
yeah my brain farted. I read it like the golem lathe could siphon ores from the station silo

**Comment by Supermichael777** [comment]
This is a bad idea because it lets nerds stay in the armored box all day. make them walk the halls so they can get dsworded with the rest.

**Comment by obscolene** [comment]
>be miner
>be first person to put materials in ORM
>go to make mining equipment from the cargo lathe
>some dickhead used all the minerals on dumb shit after hearing the mineral beep
tfw

**Comment by Wilchenx** [comment]
>be miner
>be first person to put materials in ORM
>go to make mining equipment from the cargo lathe
>some dickhead didn't research mining equipement

fixed that for you

**Comment by optimumtact** [comment]
Miners can smelt using the smelting setup if the silo is unavailable, or use a welder if that is also somehow destroyed.

**Comment by SpaceManiac** [comment]
Ready for review/testmerge/bikeshedding

**Review by AnturK — DISMISSED** [review]
Looks good. This is sheets though so testmerge first.

File: code/modules/mining/machine_silo.dm [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+	return ui.Join()
+
+/obj/machinery/ore_silo/Topic(href, href_list)
+	if(..())
+		return
+	add_fingerprint(usr)
+	usr.set_machine(src)
+
+	if(href_list["remove_orm"])
+		var/obj/machinery/mineral/orm = locate(href_list["remove_orm"])
```
Comment (optimumtact) [inline code/modules/mining/machine_silo.dm]: is there no list you can scope these locates too?

Comment (optimumtact) [reply code/modules/mining/machine_silo.dm]: wait use the orms list

File: code/modules/mining/machine_silo.dm [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+		if (istype(orm))
+			orms -= orm
+			_disconnect(orm)
+			updateUsrDialog()
+			return TRUE
+	else if(href_list["remove_lathe"])
+		var/obj/machinery/rnd/production/lathe = locate(href_list["remove_lathe"])
```
Comment (optimumtact) [inline code/modules/mining/machine_silo.dm]: use the lathes list

File: code/modules/mining/machine_silo.dm (hunk 1 above)
Comment (optimumtact) [reply code/modules/mining/machine_silo.dm]: A B S
Always Be Scoping locates

**Comment by Cyprex** [comment]
As they are, the access logs are going to become an unusable list of hundreds of items quickly, and I don't like how people can see me print ai law modules and an upload board. Looks good to me otherwise.

File: code/modules/research/machinery/_production.dm [hunk 3]
```diff
@@ -59,6 +78,11 @@
 	popup.open()
 
 /obj/machinery/rnd/production/Destroy()
+	if (silo)
+		silo.lathes -= screen
```
Comment (AnturK) [inline code/modules/research/machinery/_production.dm]: src not screen surely ?

**Review by AnturK — DISMISSED** [review]

**Comment by Wilchenx** [comment]
@SpaceManiac how about introducing a randomly generated code that you can use to link the lathes, sort of like there's a decryption key on CE's desk. It can be put on QM's desk.

Finally breaking into qm's office will be usefull for once if he doesn't secure the code.

**Comment by Tlaltecuhtli** [comment]
deconstructing and rebuilding a machine breaks the link to the commie storage which then you have to go to the vault or ask someone with access

now you need a crowbar to completly sabotage departments

**Comment by Wilchenx** [comment]
Again, introducing a password authorization system would fix that issue.

Golems could get their own silo and code, maybe even share minerals with the crew.

**Comment by Cyprex** [comment]
Lathes have to be able to pull from local storage dynamically depending on whether or not a silo is connected in my opinion. Sending local storage contents to the silo upon connecting a silo is fine if you don't want to pull from two different stocks. If lathes cannot pull from local storage sabotage becomes too easy(screwdriver crowbar shit and good luck getting through to the QM to link your new lathe and such.). On that matter, is it possible to NOT link an ORM to a silo? Imagine station crew building a silo and stealing ores because their ORM automatically sends them over.

**Comment by zxaber** [comment]
This was test-merged on Bagil this morning. Gotta say, it's a nice buff to Science. Before, there was a chance some pleb would get to the ORM first and take a sheet of something. But now we can just pull everything from the protolathe, and I don't even have to risk running into a (non-science) traitor anymore.

For real, though, being able to pull from the lathes is going to lead to departments (or greytiders) draining the vault and leaving everyone else out to dry. At the very least, it won't happen any less than the old system allowed. On the other hand, destroying or disrupting the silo wrecks Science's ability to get anything done, since the lathe can't store anything now.

So, if you don't mind ideas;
What if you gave all lathes back their storage, and had the silo fill them evenly until full. For spitball values here, 200 metal sheets, 100 glass, and a stack of 50 for all other materials. At first, all departments would get an even split. As the round progressed, departments that don't use their lathe that often would see it reach max capacity, resulting in a larger cut of the total materials being offered to those that do. And if all lathes reach max capacity, the overflow would stay in the silo until a lathe had space. This would mean that even if Science is being greedy, everyone else still has a basic cache of materials to get their cool gadgets made, and would also mean that an assistant sneaking up to the service lathe can't cripple Science.

Also, if we _are_ going to move to this model, can the Robo exosuit fabs get the same treatment?

**Comment by Bluespace** [comment]
just print out materials in science for the robotic fabs

**Comment by Wilchenx** [comment]
>exosuit fabs don't get shared minerals

👎

**Comment by Bluespace** [comment]
you know the moment exosuit fabs get shared materials they're going to print a million mech parts and never build ANY

**Comment by zxaber** [comment]
After thinking it over, I'm more glad they don't. I can at least count on mats in the exosuit fabs not getting siphoned away remotely. What I put in is what I have.

**Comment by Bluespace** [comment]
is there any way to make another silo or once it's destroyed am i just fucked for printing stuff

**Comment by Tlaltecuhtli** [comment]
the fact that machines cant store without a silo is dumb, example: greytider mc griffer breaks the silo round start, you cant print a new circuit because the printer needs glass so you are unable to use  any lathe and you have to call the shuttle,
make it so when the silo is broke or doesnt exist all minerals go back into the ORM and machines work as they did before

**Comment by blobbernaut** [comment]
Maybe make devices that use the silo require a bluespace crystal or something?

**Comment by fonedsky** [comment]
This breaks golem ship protolathes. 
image
Causes this to show up and is practically unfixable. If there was a way to make golem ship protolathes ignore silos altogether and use local material storage as they did before, that would be great. Same would apply for any normal built protolathes.
As people have pointed out here before, this is an issue with machines no longer even accepting local material storage which fucks EVERYTHING up if the silo is destroyed.

**Comment by MrDoomBringer** [comment]
image
im attempting to make a sprite for this but ill make a separate PR for it so we dont delay this one!

**Comment by Cyprex** [comment]
Thanks for your continued work on this SpaceManiac!

**Comment by 81Denton** [comment]
It looks like allow_standalone isn't working for unconnected lathes yet; last round on Bagil had people complaining that they couldn't load autolathes/protolathes on the ancient station space ruin.

**Comment by SpaceManiac** [comment]
@MrDoomBringer You asked on IRC about effects; I'm thinking one ~5ds animation which can be `flick()`ed each time there is an activity. If you don't want to wait until later to get the sprites in, you can try submitting a PR to this PR.

@81Denton The testmerge on Bagil has only just now been updated.

**Comment by GravitronBlue** [comment]
Another issue during the testmerge has been unlinked ORMs not dropping their mineral contents when deconstructed like before.

**Comment by SpaceManiac** [comment]
Done unless the testmerge surfaces any more bugs or oranges has any more requests.

**Pull request merged**